A better security-wise style bot GH Action

hanouticelina opened the pull request:
following @glegendre01's feedback [here](https://huggingface.slack.com/archives/C021H1P1HKR/p1741166002260369?thread_ts=1740109063.199639&cid=C021H1P1HKR) (private), this PR addresses the following points:

> I think the style bot action need to be in a dedicated repo to completely be secure (currently anybody can edit setup.py on the head branch). Could it be like docbuilder ?

Instead, I added a validation step at the beginning of the workflow to verify that `setup.py` and `Makefile` haven't been modified. Since the following steps only involve running `pip install` with extras and executing `make` commands, these appear to be the main "protected" files. 

> Did you test the context.payload.comment.user.login even on worklow trigger by dispatch ? (just want to be sure GH is passing correctly the context....)

I added the possibility to run the workflow manually in https://github.com/huggingface/huggingface_hub/commit/1831f6699f1b5bbb61b9964744b58f830eb34f1f. you can see the context being correctly retrieved in this [example run](https://github.com/hanouticelina/diffusers/actions/runs/13724772509). However, if i understand correctly, the context structure is different between the two trigger types. For `workflow_dispatch`, we use `context.actor`, while for `issue_comment`, we use `context.payload.comment.user.login`.

I'm not sure if the manual trigger is useful, I can remove it if not.

> I'm not very comfortable with input style_command. We can run random code, and it's not a good practice

the ability to run arbitrary style commands has been removed in favor of hardcorded commands (`make style && make quality`,`make style`, `make quality`) controlled through a case switch.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2914). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sayakpaul commented on .github/workflows/style-bot-action.yml:
```diff
@@ -45,7 +42,14 @@ jobs:
         with:
           script: |
-            const comment_user = context.payload.comment.user.login;
+            let comment_user;
+            if (context.eventName === 'workflow_dispatch') {
+              comment_user = context.actor;
+              console.log('Workflow triggered manually by:', comment_user);
+            } else {
+              comment_user = context.payload.comment.user.login;
+              console.log('Workflow triggered by comment from:', comment_user);
+            }
```
Should this only be reserved for users having admin privileges?

hanouticelina replied:
yes it's already checked here :https://github.com/huggingface/huggingface_hub/blob/3fc12e32c90b75660b4e8a6340cfa9f93abc6dd1/.github/workflows/style-bot-action.yml#L65

hanouticelina replied:
(both for workflow_dispatch and issue comment)

ydshieh replied:
well, I am not sure the details, but I kind feel `admin` is not guaranteed to every maintainer if we are talking about HF repositories.

For example, I **think** `admin` is not guaranteed to say @molbap @qubvel etc. I am `admin` because I need to access `transformers repository setting page`. 

It might be a good idea to double check here.

ydshieh replied:
If my statement above is correct, what I did so far is a lazy approach by a hard coded list of user name .... (but that is not good neither for security as people may leave the team someday)

https://github.com/huggingface/transformers/blob/2c2495cc7b0e3e2942a9310f61548f40a2bc8425/.github/workflows/self-comment-ci.yml#L32

sayakpaul commented on .github/workflows/style-bot-action.yml:
```diff
@@ -101,6 +116,33 @@ jobs:
+            const modifiedFiles = pr.map(file => file.filename);
+            console.log("Modified files:", modifiedFiles);
+            
+            const protectedFiles = ["setup.py", "Makefile"];
```
We check for this too:
https://raw.githubusercontent.com/huggingface/diffusers/refs/heads/main/utils/check_doc_toc.py

hanouticelina replied:
out of curiosity, why do you check this file in the workflow? I mean `Makefile` and `setup.py` are checked because they directly affect command execution so I wonder why `check_doc_toc.py` is also protected here.

quickly checked other HF repos, it seems that this file is not common in other repos (except for `transformers`), maybe we can add a parameter `additional_protected_files` so that you can pass `utils/check_doc_toc.py` as well. wdyt?

sayakpaul replied:
`check_doc_toc.py` is a part of the `make style && make quality` process. So, in the case it was modified, it exposes security risks.

Wauplin replied:
instead of checking for protected files, what do you think of allowing the process to run only if the  `@bot /style` comment has been written by a maintainer/reviewer? Since this bot is mostly there to help reviewers with stuck PRs, I don't think it's that much of a problem is we forbid external contributors to run it (we can always do it for them). And on the contrary it makes things more secure since it always requires a pair of eyes to validate.

To make the logic simple, I'd suggest adding a `maintainers` parameter to the GH workflow:

```yaml
# example for huggingface_hub
jobs:
  style:
    uses: ./.github/workflows/style-bot-action.yml
    with:
      python_quality_dependencies: "[quality]"
      style_command_type: "style_only"
      pr_number: ${{ fromJSON(inputs.pr_number || '0') }}
    secrets:
      bot_token: ${{ secrets.GITHUB_TOKEN }}
    maintainers: hanouticelina wauplin julien-c etc.
```

sayakpaul replied:
Oh I thought this was already implemented. In https://github.com/huggingface/diffusers/pull/10931, Dhruv from our team had already implemented it. 

But accidentally, if the `@bot /style` command is triggered somehow we still have the risk. So, I would prefer to have the modification check even if there's complexity.

hanouticelina replied:
@sayakpaul, it's indeed already implemented, see: https://github.com/huggingface/huggingface_hub/blob/4d42d1b8ba57565b9a69cc1f49920de8dd0ab6d8/.github/workflows/style-bot-action.yml#L58

Wauplin replied:
(Note that this was only my 2c, I'm not security expert here. I just find that the filter _might_ lower attention without completely removing all the attack scope. And that requiring CI approval for "make style" is quite similar to requiring CI approval for the tests suite. Both can execute malicious code if inadvertently approved. "half responsibility" is not the best way to phrase it but didn't find a better terminology :grimacing:)

hanouticelina replied:
> requiring CI approval for "make style" is quite similar to requiring CI approval for the tests suite. Both can execute malicious code if inadvertently approved.

agree! 👍

sayakpaul reviewed (commented):
Thanks! Left some comments.

ydshieh commented:
I also agree @Wauplin 's points: I'm afraid of the notion of a list of protected files is not scalable, if I look at the [Makefile in transformers](https://github.com/huggingface/transformers/blob/main/Makefile), and it's likely to miss some file in the future.

Ensure only maintainers could triggering the workflow makes more sense to me.

And a nit : `fromJSON(inputs.pr_number || '0'` doesn't make a lot of sense to me if I understand correctly. It should be a required field.

hanouticelina commented:
@ydshieh I addressed your comment in [3fc12e3](https://github.com/huggingface/huggingface_hub/pull/2914/commits/3fc12e32c90b75660b4e8a6340cfa9f93abc6dd1) 

I removed the "protected" files checking step as discussed [here](https://github.com/huggingface/huggingface_hub/pull/2914#discussion_r1987095968), the workflow is run only if the @bot /style comment has been written or triggered manually by an **admin**. Let's wait for @glegendre01 review and opinion on this and then we will be good to merge.

hanouticelina commented on .github/workflows/style-bot-action.yml:
```diff
@@ -45,14 +44,16 @@ jobs:
-            const authorized = permission.permission === 'admin';
-            console.log(`User ${comment_user} has permission level: ${permission.permission}, authorized: ${authorized} (only admins allowed)`);
+            
+            const authorized = ['admin', 'maintain', 'push'].includes(permission.permission);
+            console.log(`User ${comment_user} has permission level: ${permission.permission}, authorized: ${authorized} (only users with at least write access are allowed to run this action)`);
```
I added directly the changes from #2939 here and i will close the other PR

ydshieh commented on .github/workflows/style-bot-action.yml:
```diff
@@ -112,16 +100,25 @@ jobs:
+            if [ $COMMENT_TIMESTAMP -le $PR_MERGE_COMMIT_TIMESTAMP ]; then
+              echo "❌ Last commit on the pull request is newer than the issue comment triggering this run! Abort!";
+              exit -1;
+            fi
```
after this, we have to checkout back to the 

```
          # Instead of checking out the base repo, use the contributor's repo name
          repository: ${{ env.HEADREPOFULLNAME }}
          ref: ${{ env.HEADREF }}
```
if I understand correctly.

ydshieh replied:
(and if the above is true)

And because of this time of switching from HF's repo (with merge commit) to the contributor's repo., there is also a possibility of attack by pushing new commit after the security check is finished.

I can provide some possible options to avoid this, if you can confirm we do want to checkout back to contributor's repo.

hanouticelina replied:
yeah I think you're right, we need to checkout the PR's head branch again. what do you suggest to avoid any attack in that case?

hanouticelina replied:
okay I added another checkout with the same `HEADREF` after the timestamp check step in [3b03d35](https://github.com/huggingface/huggingface_hub/pull/2914/commits/3b03d353504c319b30fb73e55374163ae67bad2f).

ydshieh replied:
(sorry, my previous message might be a bit misleading , as I didn't think it thoroughly) 

I think it could be simple as just `git checkout $HEADREF`.

From the first place where we checkout with

```
        uses: actions/checkout@v3
        env: 	        env: 
          HEADREPOFULLNAME: ${{ steps.pr_info.outputs.headRepoFullName }}	          HEADREPOFULLNAME: ${{ steps.pr_info.outputs.headRepoFullName }}
```
the `origin` is `headRepoFullName `. Later when we fetch/checkout , it doesn't change the `origin` although it checkout to the content of merge commit from `refs/pull/$PR_NUMBER/merge`.

So later when we do `git push origin HEAD:$HEADREF`, the origin is still the contributor's repo.

If this works, we don't need an extra step. WDYT?

hanouticelina replied:
addressed in [1063565](https://github.com/huggingface/huggingface_hub/pull/2914/commits/106356502f26efd4a3c166f8f2307fb12d46c981)

sayakpaul commented:
Okay if I review this after I am back from my vacation (6th April).

hanouticelina commented:
merging! (failing tests are unrelated)
